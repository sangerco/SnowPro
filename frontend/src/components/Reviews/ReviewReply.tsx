import React, { useEffect } from 'react';
import { fetchNewReviewReplyDataByReviewId } from '../../redux/actions/replyReviewActions';
import { connect } from 'react-redux';
import { RootState } from '../../redux/store';
import { Container, Header } from 'semantic-ui-react';

interface ReviewReplyData {
    id: string;
    reviewId: string;
    userId: string;
    username: string;
    body: string,
    slug: string,
    createdAt: Date;
};

interface ReviewReplyProps {
    id: string;
    slug: string;
    reviewReply: ReviewReplyData | null;
    error: string | null;
    fetchNewReviewReplyDataByReviewId: (slug: string, id: string) => void;
}

const ReviewReply: React.FC<ReviewReplyProps> = ({ id, slug, reviewReply, error, fetchNewReviewReplyDataByReviewId }) => {
    useEffect(() => {
        fetchNewReviewReplyDataByReviewId(id, slug);
    }, [id, slug, fetchNewReviewReplyDataByReviewId]);

    const formatDate = (date: Date) => {
        const newDate = new Date(date);
        const options: Intl.DateTimeFormatOptions = { year: 'numeric', month: 'long', day: 'numeric' };
        return newDate.toLocaleDateString('en-US', options);
    }

    if(error) return `Error retrieving this reply! ${error}`;

    if(reviewReply) {
        const date = formatDate(reviewReply.createdAt)

        return (
            <div>
                <Container textAlign='left'><Header as='h6'>By {reviewReply.username}</Header></Container>
                <Container textAlign='right'><i>{date}</i></Container>
                <Container text>
                    <p>{reviewReply.body}</p>
                </Container>
            </div>
        );
    };

    return null;
};

const mapStateToProps = (state: RootState) => ({
    reply: state.reviewReply.data,
    error: state.reviewReply.error
});

const mapDispatchToProps = {
    fetchNewReviewReplyDataByReviewId
};

export default connect(mapStateToProps, mapDispatchToProps)(ReviewReply);