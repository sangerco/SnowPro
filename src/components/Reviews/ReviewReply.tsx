import React, { useState, useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { RootState, AppDispatch } from "../../redux/store";
import { Link, useParams } from "react-router-dom";
import { fetchReview } from "../../redux/slices/reviewSlice";
import {
  fetchReviewReply,
  deleteReviewReply,
} from "../../redux/slices/reviewReplySlice";
import {
  Container,
  Dimmer,
  Divider,
  Header,
  Loader,
  Icon,
  Modal,
  Button,
} from "semantic-ui-react";

const ReviewReply: React.FC = () => {
  const auth = useSelector((state: RootState) => state.auth);
  const authId = auth.data?.id;
  const dispatch = useDispatch<AppDispatch>();
  const { id } = useParams();

  const [showDeleteModal, setShowDeleteModal] = useState(false);

  useEffect(() => {
    if (id) {
      dispatch(fetchReviewReply(id));
    }
  }, [dispatch, id]);

  const reviewReplies = useSelector((state: RootState) => state.reviewReplies);
  const reviewReply = reviewReplies.reviewReply;

  const reviewId = reviewReply?.reviewId;

  useEffect(() => {
    if (reviewId) {
      dispatch(fetchReview(reviewId));
    }
  }, [dispatch, reviewId]);

  const reviews = useSelector((state: RootState) => state.reviews);
  const review = reviews.review;

  const handleShowDeleteModal = () => {
    setShowDeleteModal(true);
  };

  const handleDeleteReviewReply = () => {
    if (id) {
      dispatch(deleteReviewReply(id));
    }
  };

  const formatDate = (date: Date) => {
    const newDate = new Date(date);
    const options: Intl.DateTimeFormatOptions = {
      year: "numeric",
      month: "long",
      day: "numeric",
    };
    return newDate.toLocaleDateString("en-US", options);
  };

  if (reviewReplies.error) {
    return (
      <Dimmer active>
        <div>
          <p>
            Error! Review Reply cannot be retrieved! {`${reviewReplies.error}`}
          </p>
        </div>
      </Dimmer>
    );
  }

  if (reviewReplies.loading) {
    return (
      <Dimmer active>
        <Loader>Loading...</Loader>
      </Dimmer>
    );
  }

  if (Array.isArray(reviewReply) || Array.isArray(review)) {
    <Dimmer active>
      <p>Error! Review Reply cannot be retrieved!</p>
    </Dimmer>;
  } else if (reviewReply && review) {
    const date = formatDate(reviewReply.createdAt);
    const userId = reviewReply.userId;

    return (
      <div>
        <Container fluid>
          <Header as="h3">
            Reply to{" "}
            {
              <Link to={`/ski-areas/reviews/${review.id}`}>
                {review.header}
              </Link>
            }
          </Header>
          <Divider />
          <Header as="h4">By {reviewReply.username}</Header>
          <Header textAlign="right" as="h6">
            {date}
          </Header>
          <Divider />
          <p>{reviewReply.body}</p>
          {authId && authId === userId ? (
            <Container>
              <Link to={`/ski-areas/reviews/replies/${reviewReply.id}/update`}>
                <Icon name="edit" style={{ cursor: "pointer" }} />
              </Link>
              <Icon
                name="trash"
                style={{ cursor: "pointer" }}
                onClick={handleShowDeleteModal}
              />
            </Container>
          ) : null}
        </Container>
        <Modal open={showDeleteModal} onClose={() => setShowDeleteModal(false)}>
          <Modal.Content>
            Are You Sure You Want To Delete This Review?
          </Modal.Content>
          <Modal.Actions>
            <Button negative onClick={handleDeleteReviewReply}>
              Yes
            </Button>
            <Button onClick={() => setShowDeleteModal(false)}>Cancel</Button>
          </Modal.Actions>
        </Modal>
      </div>
    );
  }

  return null;
};

export default ReviewReply;
