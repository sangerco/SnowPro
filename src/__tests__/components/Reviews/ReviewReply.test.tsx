/* eslint-disable testing-library/no-node-access */
import React from "react";
import ReviewReply from "../../../components/Reviews/ReviewReply";
import { render, fireEvent, screen, waitFor } from "@testing-library/react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import nock, { pendingMocks } from "nock";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

describe("test Review Reply", () => {
  const initialState = {
    auth: {
      data: {
        id: "11",
      },
    },
    reviews: {
      reviews: [
        {
          id: "review-1",
          userId: "11",
          username: "testuser",
          skiAreaSlug: "ski-area-1",
          skiAreaName: "Ski Area 1",
          header: "test review",
          body: "test review body",
          stars: 5,
          photos: [],
          createdAt: "fake created date",
          replies: ["review-reply-1"],
        },
        {
          id: "review-2",
          userId: "22",
          username: "testuser2",
          skiAreaSlug: "ski-area-2",
          skiAreaName: "Ski Area 2",
          header: "test review 2",
          body: "test review body 2",
          stars: 1,
          photos: [],
          createdAt: "fake created date",
          replies: [],
        },
      ],
    },
    reviewReplies: {
      reviewReply: {
        id: "review-reply-1",
        reviewId: "review-1",
        userId: "22",
        username: "testuser2",
        body: "test review reply body",
        slug: "ski-area-1",
        createdAt: "fake created date",
      },
    },
    users: {
      users: [
        {
          userId: "22",
          username: "testuser2",
        },
        {
          userId: "33",
          username: "testuser3",
        },
      ],
    },
    skiAreas: {
      skiAreas: [
        {
          slug: "ski-area-1",
          name: "Ski Area 1",
          country: "US",
          region: "CO",
          href: "www.fakehref.com",
          location: {
            latitude: 0,
            longitude: 0,
          },
          lifts: {},
          twitter: {},
        },
        {
          slug: "ski-area-1",
          name: "Ski Area 1",
          country: "US",
          region: "CO",
          href: "www.fakehref.com",
          location: {
            latitude: 0,
            longitude: 0,
          },
          lifts: {},
          twitter: {},
        },
      ],
    },
  };
  const store = mockStore(initialState);
  it("should render without crashing", () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <ReviewReply />
        </BrowserRouter>
      </Provider>
    );
  });
  it("should match snapshot", () => {
    const { asFragment } = render(
      <Provider store={store}>
        <BrowserRouter>
          <ReviewReply />
        </BrowserRouter>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useParams: jest.fn(),
}));

describe("test review reply view", () => {
  const id = "review-reply-1";
  const useParamsMock = jest.requireMock("react-router-dom")
    .useParams as jest.Mock;
  useParamsMock.mockReturnValue({ id });
  const initialState = {
    auth: {
      data: {
        id: "22",
      },
    },
    reviews: {
      review: {
        id: "review-1",
        userId: "11",
        username: "testuser",
        skiAreaSlug: "ski-area-1",
        skiAreaName: "Ski Area 1",
        header: "test review",
        body: "test review body",
        stars: 5,
        photos: [],
        createdAt: "fake created date",
        replies: ["review-reply-1"],
      },
      reviews: [
        {
          id: "review-1",
          userId: "11",
          username: "testuser",
          skiAreaSlug: "ski-area-1",
          skiAreaName: "Ski Area 1",
          header: "test review",
          body: "test review body",
          stars: 5,
          photos: [],
          createdAt: "fake created date",
          replies: ["review-reply-1"],
        },
        {
          id: "review-2",
          userId: "22",
          username: "testuser2",
          skiAreaSlug: "ski-area-2",
          skiAreaName: "Ski Area 2",
          header: "test review 2",
          body: "test review body 2",
          stars: 1,
          photos: [],
          createdAt: "fake created date",
          replies: [],
        },
      ],
    },
    reviewReplies: {
      reviewReply: {
        id: "review-reply-1",
        reviewId: "review-1",
        userId: "22",
        username: "testuser2",
        body: "test review reply body",
        slug: "ski-area-1",
        createdAt: "fake created date",
      },
    },
    users: {
      users: [
        {
          userId: "22",
          username: "testuser2",
        },
        {
          userId: "33",
          username: "testuser3",
        },
      ],
    },
    skiAreas: {
      skiAreas: [
        {
          slug: "ski-area-1",
          name: "Ski Area 1",
          country: "US",
          region: "CO",
          href: "www.fakehref.com",
          location: {
            latitude: 0,
            longitude: 0,
          },
          lifts: {},
          twitter: {},
        },
        {
          slug: "ski-area-1",
          name: "Ski Area 1",
          country: "US",
          region: "CO",
          href: "www.fakehref.com",
          location: {
            latitude: 0,
            longitude: 0,
          },
          lifts: {},
          twitter: {},
        },
      ],
    },
  };
  const store = mockStore(initialState);
  it("should display a review reply", async () => {
    nock("http://localhost:5000")
      .defaultReplyHeaders({
        "access-control-allow-origin": "*",
      })
      .get("/ski-areas/reviews/replies/review-reply-1")
      .reply(200, {
        id: "review-reply-1",
        reviewId: "review-1",
        userId: "22",
        username: "testuser2",
        body: "test review reply body",
        slug: "ski-area-1",
        createdAt: "fake created date",
      });
    nock("http://localhost:5000")
      .defaultReplyHeaders({
        "access-control-allow-origin": "*",
      })
      .get(`/ski-areas/reviews/review-1`)
      .reply(200, {
        id: "review-1",
        userId: "11",
        username: "testuser",
        skiAreaSlug: "ski-area-1",
        skiAreaName: "Ski Area 1",
        header: "test review",
        body: "test review body",
        stars: 5,
        photos: [],
        createdAt: "fake created date",
        replies: [],
      });
    nock("http://localhost:5000")
      .defaultReplyHeaders({
        "access-control-allow-origin": "*",
      })
      .get("/ski-areas")
      .reply(200, []);
    render(
      <Provider store={store}>
        <BrowserRouter>
          <ReviewReply />
        </BrowserRouter>
      </Provider>
    );

    await waitFor(() => {
      const reviewReply = screen.getByTestId("review-reply");
      expect(reviewReply).toBeInTheDocument();
    });

    await waitFor(() => {
      const buttons = screen.getByTestId("buttons");
      expect(buttons).toBeInTheDocument();
    });
  });

  it("should display a modal", async () => {
    nock("http://localhost:5000")
      .defaultReplyHeaders({
        "access-control-allow-origin": "*",
      })
      .get("/ski-areas/reviews/replies/review-reply-1")
      .reply(200, {
        id: "review-reply-1",
        reviewId: "review-1",
        userId: "22",
        username: "testuser2",
        body: "test review reply body",
        slug: "ski-area-1",
        createdAt: "fake created date",
      });
    nock("http://localhost:5000")
      .defaultReplyHeaders({
        "access-control-allow-origin": "*",
      })
      .get(`/ski-areas/reviews/review-1`)
      .reply(200, {
        id: "review-1",
        userId: "11",
        username: "testuser",
        skiAreaSlug: "ski-area-1",
        skiAreaName: "Ski Area 1",
        header: "test review",
        body: "test review body",
        stars: 5,
        photos: [],
        createdAt: "fake created date",
        replies: [],
      });
    nock("http://localhost:5000")
      .defaultReplyHeaders({
        "access-control-allow-origin": "*",
      })
      .get("/ski-areas")
      .reply(200, []);
    render(
      <Provider store={store}>
        <BrowserRouter>
          <ReviewReply />
        </BrowserRouter>
      </Provider>
    );

    const deleteButton = screen.getByTestId("delete-button");
    fireEvent.click(deleteButton);

    await waitFor(() => {
      const modal = screen.getByTestId("delete-modal");
      expect(modal).toBeInTheDocument();
    });
  });
});
