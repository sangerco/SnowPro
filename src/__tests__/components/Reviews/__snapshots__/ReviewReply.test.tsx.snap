// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Review Reply should match snapshot 1`] = `<DocumentFragment />`;
